skbio/embedding/_protein.py

```diff
@@ -0,0 +1,90 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+from skbio.sequence import Protein
+from skbio.embedding._embedding import SequenceEmbedding
+from skbio.util import get_data_path
+from pathlib import Path
+
+
+class ProteinEmbedding(SequenceEmbedding):
+    """Stores the embeddings of the protein sequence.
+
+    Parameters
+    ----------
+    sequence : str, Sequence, or 1D np.ndarray
+        Characters representing the protein sequence itself.
+    embedding : np.ndarray
+        The embedding of the protein sequence. Row vectors correspond to
+        the latent residues coordinates.
+    clip_head : bool, optional
+        If ``True``, then the first row of the embedding will be removed.
+        Some language models specify start tokens, and this parameter can
+        be used to account for this.
+    clip_end : bool, optional
+        If ``True``, then the last row of the embedding will be removed.
+        Some language models specify end tokens, and this parameter can
+        be used to account for this.
+
+    See Also
+    --------
+    Protein
+
+    """
+
+    default_write_format = "embed"
+
+    def __init__(
+        self, embedding, sequence: str, clip_head=False, clip_tail=False, **kwargs
+    ):
+        if clip_head:
+            embedding = embedding[1:]
+        if clip_tail:
+            embedding = embedding[:-1]
+
+        if isinstance(sequence, Protein):
+            sequence = str(sequence)
+
+        # make sure that the embedding has the same length as the sequence
+        sequence_len = len(sequence)
+        if embedding.shape[0] != sequence_len:
+            raise ValueError(
+                f"The embedding ({embedding.shape[0]}) must have the "
+                f"same length as the sequence ({len(sequence)})."
+            )
+
+        super(ProteinEmbedding, self).__init__(
+            embedding=embedding, sequence=sequence, **kwargs
+        )
+
+    def __str__(self):
+        return str(self._ids)
```

Comment:
This looks good, but when I do functional tests and I do `test_object.__str__` the output is identical to `test_object.__repr__`. Is this correct? It looks like `test_object.__str__` should just give me a string of the sequence(s). If I do `str(test_object._ids` I get the expected output of a string.

Reply:
Could you clarify this with a minimal example? There is a unittest for the repr that seems to be passing on my end.

Reply:
Sure, the repr looks good to me. I am just confused whether the `__str__` is outputting what it should. 

````python
>>> test = ProteinEmbedding(sequence=sequence, embedding=embeddings, clip_head=True, clip_tail=True)
>>> test.__repr__
... <bound method ProteinEmbedding.__repr__ of ProteinEmbedding
... -------------------------
... Stats:
...     length: 7
...     embedding dimension: 1024
...     has gaps: False
...     has degenerates: True
...     has definites: True
...     has stops: False
... -------------------------
... 0 AETCXAX>
>>> test.__str__
... <bound method ProteinEmbedding.__repr__ of ProteinEmbedding
... -------------------------
... Stats:
...     length: 7
...     embedding dimension: 1024
...     has gaps: False
...     has degenerates: True
...     has definites: True
...     has stops: False
... -------------------------

I'm not sure if there's anything wrong with this. I'm just curious if `__str__` and `__repr__` are supposed to output the same thing?

Reply:
````python
>>> str(test)
... 'AETCXAX'
>>> print(repr(test))
... ProteinEmbedding
... -------------------------
... Stats:
...     length: 7
...     embedding dimension: 1024
...     has gaps: False
...     has degenerates: True
...     has definites: True
...     has stops: False
... -------------------------
... 0 AETCXAX
````
So if done with `repr(test)` and `str(test)` it works as intended.

Reply:
Awesome awesome :) so in case this wasn't known, many (not all) of the dunders have functions that access like these, but also see `hash`, `len`, etc. And `==` for `__eq__`, etc